content/doc/book/pipeline/development.adoc

```diff
@@ -0,0 +1,145 @@
+
+[[linter]]
+== Command-line Pipeline Linter (Declarative only)
+
+Jenkins can validate, or
+"link:https://en.wikipedia.org/wiki/Lint_(software)[lint]",
+a Declarative Pipleline from the command line before running actually running it.
+This can be done using a Jenkins CLI command or using `curl` to post the Jenkinsfile directly.
+We recommended using the
+<<../managing/cli#ssh, SSH interface>>
+for running the linter.
+
+NOTE: See the <<../managing/cli, Jenkins CLI documentation>> for details on how to properly configure
+Jenkins for secure command-line access.
+
+// link:https://www.owasp.org/index.php/Cross-Site_Request_Forgery_(CSRF)[CRSF] protection
+
+[source,bash]
+----
+# ssh (Jenkins CLI)
+# JENKINS_SSHD_PORT=[sshd port on master]
+# JENKINS_HOSTNAME=[Jenkins master hostname]
+ssh -p $JENKINS_SSHD_PORT $JENKINS_HOSTNAME declarative-linter < Jenkinsfile
+
+# curl (REST API)
+# Assuming "anonymous read access" has been enabled on your Jenkins instance.
+# JENKINS_URL=[root URL of Jenkins master]
+# JENKINS_CRUMB is needed if your Jenkins master has CRSF protection enabled as it should
+JENKINS_CRUMB=`curl "$JENKINS_URL/crumbIssuer/api/xml?xpath=concat(//crumbRequestField,\":\",//crumb)"`
+curl -X POST -H $JENKINS_CRUMB -F "jenkinsfile=<Jenkinsfile" $JENKINS_URL/pipeline-model-converter/validate
+----
+
+=== Examples
+
+.Invalid Jenkinsfile
+[source,bash]
+----
+# pass a Jenkinsfile that does not contain an "agent" section
+ssh -p 8675 localhost declarative-linter < ./Jenkinsfile
+Errors encountered validating Jenkinsfile:
+WorkflowScript: 25: Not a valid section definition: 'agent'. Some extra configuration is required. @ line 25, column 1.
+   pipeline &#125;
+   ^
+
+WorkflowScript: 25: Missing required section 'agent' @ line 25, column 1.
+   pipeline &#125;
+   ^
+----
+
+.Valid Jenkinsfile
+[source,bash]
+----
+ssh -p 8675 localhost declarative-linter < ./Jenkinsfile
+Jenkinsfile successfully validated.
+----
+
+[[replay]]
+== "Replay" Pipeline Runs with Modifications
+
+Creating Pipelines and debugging failures in Pipelines can be difficult and time consuming.
+The standard development process for Pipeline changes -
+change, commit, push, run, and analyze - add quite a bit of overhead between iterations.
+Sometimes we just want to try out a quick change to a Pipeline and see the result.
+
+The "Replay" feature lets us quickly test modifications to a Pipeline
+without creating a new commit or following regular code creation process.
+To use the "Replay" feature:
+
+. Select a previously completed run in the build history.
+image::pipeline/replay-previous-run.png[Previous Pipeline Run]
+. Click "Replay" in the left menu
+image::pipeline/replay-left-bar.png[Replay Left-menu Button]
+. Make modifications and click "Run".
+In this example, we changed "ruby-2.3" to "ruby-2.4".
+image::pipeline/replay-modified.png[Replay Left-menu Button]
+. Check the results of changes
+
+Replay can be called multiple times on the same run,
```

Comment:
I think another paragraph should be added about Replay allowing the user to make changes to Shared Libraries loaded for the Pipeline as well.